PhysicsTools/HepMCCandAlgos/plugins/GenPUProtonProducer.cc

```diff
@@ -0,0 +1,255 @@
+/* \class GenPUProtonProducer
+ *
+ * Modification of GenParticleProducer.
+ * Saves final state protons from HepMC events in Crossing Frame, in the generator-particle format.
+ *
+ * Note: Use the option USER_CXXFLAGS=-DEDM_ML_DEBUG with SCRAM in order to enable debug messages.
+ *
+ *    March  9, 2017   : Initial version.
+ *    March 14, 2017   : Updated debug messages.
+ *     July 27, 2017   : Removed extra loop initially inherited from GenParticleProducer. 
+ *   August 17, 2017   : Replaced std::auto_ptr with std::unique_ptr. 
+ * September 6, 2017   : Updated module to edm::global::EDProducer with ConvertParticle as RunCache following GenParticleProducer. 
+ *
+ */
+
+#include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
+#include "SimGeneral/HepPDTRecord/interface/ParticleDataTable.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+#include "FWCore/Utilities/interface/EDMException.h"
+
+#include <vector>
+#include <string>
+#include <unordered_map>
+
+namespace {
+
+  class ConvertParticle {
+    public:
+     static constexpr int PDGCacheMax = 32768;
+     static constexpr double mmToCm = 0.1;
+
+     ConvertParticle() :
+        abortOnUnknownPDGCode_( true ),
+        initialized_( false ),
+        chargeP_( PDGCacheMax, 0 ), chargeM_( PDGCacheMax, 0 ) {};
+
+     ConvertParticle(bool abortOnUnknownPDGCode) :
+        abortOnUnknownPDGCode_( abortOnUnknownPDGCode ),
+        initialized_( false ),
+        chargeP_( PDGCacheMax, 0 ), chargeM_( PDGCacheMax, 0 ) {};
+
+     ~ConvertParticle() {};
+
+     bool initialized() const { return initialized_; }
+
+     void init(HepPDT::ParticleDataTable const& pdt) {
+	if( !initialized_ ) {
+	   for( HepPDT::ParticleDataTable::const_iterator p = pdt.begin(); p != pdt.end(); ++p ) {
+	      HepPDT::ParticleID const& id = p->first;
+	      int pdgId = id.pid(), apdgId = std::abs( pdgId );
+	      int q3 = id.threeCharge();
+	      if ( apdgId < PDGCacheMax && pdgId > 0 ) {
+		 chargeP_[ apdgId ] = q3;
+		 chargeM_[ apdgId ] = -q3;
+	      } else if ( apdgId < PDGCacheMax ) {
+		 chargeP_[ apdgId ] = -q3;
+		 chargeM_[ apdgId ] = q3;
+	      } else {
+	         chargeMap_.emplace( pdgId, q3);
+	         chargeMap_.emplace( -pdgId, -q3);
+	      }
+	   }
+	   initialized_ = true;
+	}
+     }
+
+     bool operator() (reco::GenParticle& cand, HepMC::GenParticle const* part) const {
+	reco::Candidate::LorentzVector p4( part->momentum() );
+	int pdgId = part->pdg_id();
+	cand.setThreeCharge( chargeTimesThree( pdgId ) );
+	cand.setPdgId( pdgId );
+	cand.setStatus( part->status() );
+	cand.setP4( p4 );
+	cand.setCollisionId(0);
+	HepMC::GenVertex const* v = part->production_vertex();
+	if ( v != nullptr ) {
+	   HepMC::ThreeVector vtx = v->point3d();
+	   reco::Candidate::Point vertex( vtx.x() * mmToCm, vtx.y() * mmToCm, vtx.z() * mmToCm );
```

Comment:
hi @antoniovilela - i'm (naively) surprised by the need for this unit conversion.. It does match a few other codes I find - but what is the source of these non-cmd-standard units

Reply:
I understand the HepMC internal unit is mm, which is converted to cm for the GenParticle format. We see later that the vertex position distribution is consistent.